tests/semantic_analyser.cpp

```diff
@@ -3006,6 +3006,254 @@ TEST_F(semantic_analyser_btf, fentry)
   test("fentry:func_1 { $x = args->a; }", 0);
 }
 
+TEST(semantic_analyser, for_loop_map_one_key)
+{
+  test("BEGIN { @map[0] = 1; for ($kv : @map) { print($kv); } }", R"(
+Program
+ BEGIN
+  =
+   map: @map :: [int64]
+    int: 0 :: [int64]
+   int: 1 :: [int64]
+  for
+   decl
+    variable: $kv :: [(unsigned int64,int64)]
+   expr
+    map: @map :: [int64]
+   stmts
+    call: print
+     variable: $kv :: [(unsigned int64,int64)]
+)");
+}
+
+TEST(semantic_analyser, for_loop_map_two_keys)
+{
+  test("BEGIN { @map[0,0] = 1; for ($kv : @map) { print($kv); } }", R"(
+Program
+ BEGIN
+  =
+   map: @map :: [int64]
+    int: 0 :: [int64]
+    int: 0 :: [int64]
+   int: 1 :: [int64]
+  for
+   decl
+    variable: $kv :: [((unsigned int64,unsigned int64),int64)]
+   expr
+    map: @map :: [int64]
+   stmts
+    call: print
+     variable: $kv :: [((unsigned int64,unsigned int64),int64)]
+)");
+}
+
+TEST(semantic_analyser, for_loop_map)
+{
+  test("BEGIN { @map[0] = 1; for ($kv : @map) { print($kv); } }");
+  test("BEGIN { @map[0] = 1; for ($kv : @map) { print($kv.0); } }");
+  test("BEGIN { @map[0] = 1; for ($kv : @map) { print($kv.1); } }");
+}
+
+TEST(semantic_analyser, for_loop_map_assign_decl_to_var)
```

Comment:
Should we add the negative test to show it errors with the wrong type.

Reply:
Like JSON output, I think variable assignments are generally orthogonal to for-loops. I can't actually remember why I added this test in the first place. Will remove it if I can't work it out.

I'd think that all we need to check is that `$kv` has the correct type, and the test case `for_loop_map_one_key` is already doing that by printing out the types in the AST.